### PR TITLE
Remove command prefixes on commandline

### LINKED ISSUE #1018: ingame help command help misleading
help

### PR DESCRIPTION
**Pull request by Descacharrado**
Commands will no longer have a prefix in the commandline, but they will keep it in the chat commands.

Closes #1018

### CODE REVIEW

**Comment by ProditorMagnus** [comment]
It does not look correct to assume prefix is always :.

**Comment by Descacharrado** [comment]
> It does not look correct to assume prefix is always :.

It does not assume prefix to be ':', it tests for a command to have ':' as a prefix. To clear things up I tested this with chat commands and game commands:

Chat commands, in order, top to bottom: (emote, :emote, /emote, /emtoe, /:emote, :/emote)
commands in chat

Game commands, in order, top to bottom: (benchmark, :benchmark, benhcmark, /benchmark, /:benchmark, :/benchmark)
commands in twodots

I may have missed somewhere you can input commands too (not just ingame by pressing : or m), in that case, tell me and I'll test those.

File: src/map_command_handler.hpp
Comment (Pentarctagon) [inline src/map_command_handler.hpp:196]: Is there a reason to declare `compare` rather than just directing checking `:`?

Comment (Descacharrado) [reply src/map_command_handler.hpp]: I though I had tried it and it gave me an error, but looks like I did something wrong then, it's now cleaner (on 3d4d406).

**Comment by Pentarctagon** [comment]
It looks good to me.

**Review by stevecotton — CHANGES_REQUESTED** [review]
This code is also used for the chat commands, for which `cmd_prefix_` is `/` and the `help_command_` is `/help` - removing the `/` will say "help" instead. I like the idea, but the implementation needs to handle these possibilities too.

**Comment by Descacharrado** [comment]
> This code is also used for the chat commands, for which `cmd_prefix_` is `/` and the `help_command_` is `/help` - removing the `/` will say "help" instead. I like the idea, but the implementation needs to handle these possibilities too.

So we would want the previous implementation for chat commands. My idea was that if the player already knew how to enter a command (either using ':' or opening the chat and using '/') they would already know that 'help' refers to /help, but it is true that it may cause confusion an make people write help in chat instead of /help. Will commit changes later.

**Comment by CelticMinstrel** [comment]
> Comparing what is written to existing commands and adding suggestions

Were you planning to do this? It's not actually that hard – just compute the Levenshtein distance between the input command and each available commands and suggest those for which the result is less than a certain threshold.

**Comment by Descacharrado** [comment]
> Were you planning to do this? It's not actually that hard – just compute the Levenshtein distance between the input command and each available commands and suggest those for which the result is less than a certain threshold.

I wasn't planning on doing it in this PR but I was searching for ways to do it, thanks for the guidance. I wouldnt wait for me to do that before merging tho, if I develop that succesfully I'll just open another PR if this one is closed.

File: src/map_command_handler.hpp
Comment (stevecotton) [inline src/map_command_handler.hpp:195]: Doing it like this still means that the class treats two prefix strings specially. Instead of hardcoding strings, please check whether `input` starts with `cmd_prefix_`. We're currently using C++17, so we don't have `std::string::starts_with()`, but we do have `boost::algorithm::starts_with()`. @CelticMinstrel do we have a utils function for that (I thought so, but can't find it)?

I think the logic should check whether get_command would succeed for the suggested change, so that `:help` would get a hint to remove the `cmd_prefix_`, but `:unknown` wouldn't.

Comment (CelticMinstrel) [reply src/map_command_handler.hpp:195]: I'm not aware of one?

Even without Boost or C++17, "starts with" is just `str.find(x) == 0`, right?

Comment (Descacharrado) [reply src/map_command_handler.hpp:195]: > Doing it like this still means that the class treats two prefix strings specially. 

Wait, isn't that what we want? We want the `"help_command"` symbol to be `cmd_prefix_ + "help"` when it's a chat command and just `"help"` when its a "normal" command, so we need to make a distinction (unless there is another way of checking whether you are in a chat command or in a "normal" command which I do not know of). Did you mean that I should be using something similar to `input[0] == cmd_prefix_` instead of `input[0] == ':'` in line 201?

> I think the logic should check whether get_command would succeed for the suggested change, so that :help would get a hint to remove the cmd_prefix_, but :unknown wouldn't.

I feel like that is reasonable.

Comment (stevecotton) [reply src/map_command_handler.hpp:195]: > Did you mean that I should be using something similar to input[0] == cmd_prefix_ instead of input[0] == ':' in line 201?

That's what I meant, although it might be multiple characters so it should be a starts-with check instead. Also, the translatable strings should put cmd_prefix_ in the symbols instead of including it in the string itself.

File: src/map_command_handler.hpp
Comment (stevecotton) [inline src/map_command_handler.hpp:206]: Something that could be left to a follow-up PR would be to convert the `'` marks to match the https://wiki.wesnoth.org/Typography_Style_Guide . There are other strings in this file that aren't involved in the current change, so I'd be happy to postpone all of these to a follow-up.

File: src/map_command_handler.hpp
Comment (Descacharrado) [reply src/map_command_handler.hpp:194]: e4506d5 should handle both requests.

File: src/map_command_handler.hpp
Comment (Descacharrado) [reply src/map_command_handler.hpp:213]: I would agree.

File: src/map_command_handler.hpp
Comment (stevecotton) [reply src/map_command_handler.hpp:194]: It still has a hardcoded `/`, and an assumption that cmd_prefix_ is exactly 1 character.

File: src/map_command_handler.hpp
Comment (stevecotton) [inline src/map_command_handler.hpp:203]: `c` isn't used, so it shouldn't need to store this to a variable.
The `erase(0,1)` is hardcoding the assumed length of `cmd_prefix`
Why use `get_data(0)` instead of `get_cmd()`?

File: src/map_command_handler.hpp
Comment (stevecotton) [inline src/map_command_handler.hpp:209]: This and the following `else` print the same message, could merge them by merging the two `if` statements that get to this line.

**Comment by Descacharrado** [comment]
Drafting until complete.

**Comment by Descacharrado** [comment]
Succesfully changed the hardcoded "/" for a flag. With 9de67a5, if the command prefix is ever changed, it will not be necessary to modify `map_command_handler.hpp`

File: src/map_command_handler.hpp
Comment (stevecotton) [inline src/map_command_handler.hpp:390]: There's no documentation about what the "command or message" flag means. Please document how this method changes the observable behavior.

Comment (Descacharrado) [reply src/map_command_handler.hpp]: Hope it is now fine. Could be shorter, but I like explaining thoroughly.

**Comment by Descacharrado** [comment]
Converting to draft until I resolve conflicts (because of my other PR).

**Comment by Descacharrado** [comment]
Uhhhhhh I just noticed an error, wait for another commit before review.

**Comment by Pentarctagon** [comment]
@stevecotton is this good to merge at this point?

**Comment by stevecotton** [comment]
I think it's good to merge, will do that tomorrow unless there are more comments.

I'd prefer if the comments were in Doxygen format and reworded, but the rest of the class isn't in Doxygen format either and at the moment I can't work out a clearer way to write the text that I want to change. All of that sounds like stuff for a later PR rather than the current one.

**Review by stevecotton — APPROVED** [review]

**Comment by Descacharrado** [comment]
> I'd prefer if the comments were in Doxygen format and reworded, but the rest of the class isn't in Doxygen format either and at the moment I can't work out a clearer way to write the text that I want to change. All of that sounds like stuff for a later PR rather than the current one.

I'll revise the format in a follow-up PR.

**Pull request merged**
